jolt-core/src/jolt/vm/ram.rs

```diff
@@ -1502,50 +1677,105 @@ fn prove_ra_hamming_weight<F: JoltField, ProofTranscript: Transcript>(
+    let log_k = K.log_2();
+    let base_chunk_size = log_k / d;
+    let remainder = log_k % d;
+    let chunk_sizes: Vec<usize> = (0..d)
+        .map(|i| {
+            if i < remainder {
+                base_chunk_size + 1
+            } else {
+                base_chunk_size
+            }
+        })
+        .collect();
+
+    // Compute F arrays for each decomposed part
+    let mut F_arrays: Vec<Vec<F>> = vec![vec![F::zero(); K]; d];
+
+    trace
         .par_chunks(chunk_size)
         .enumerate()
         .map(|(chunk_index, trace_chunk)| {
-            let mut result = unsafe_allocate_zero_vec(K);
+            let mut local_arrays: Vec<Vec<F>> = vec![unsafe_allocate_zero_vec(K); d];
             let mut j = chunk_index * chunk_size;
             for cycle in trace_chunk {
-                let k = remap_address(cycle.ram_access().address() as u64, memory_layout) as usize;
-                result[k] += eq_r_cycle[j];
+                let address =
+                    remap_address(cycle.ram_access().address() as u64, memory_layout) as usize;
+
+                // For each address, add eq_r_cycle[j] to each corresponding chunk
+                // This maintains the property that sum of all ra values for an address equals 1
+                let mut remaining_address = address;
+                let mut chunk_values = Vec::with_capacity(d);
+
+                // Decompose address into chunks
+                for i in 0..d {
+                    let chunk_size = chunk_sizes[d - 1 - i];
+                    let chunk_modulo = 1 << chunk_size;
+                    let chunk_value = remaining_address % chunk_modulo;
+                    chunk_values.push(chunk_value);
+                    remaining_address /= chunk_modulo;
+                }
+
+                // Add eq_r_cycle contribution to each ra polynomial
+                for (i, &chunk_value) in chunk_values.iter().enumerate() {
+                    local_arrays[d - 1 - i][chunk_value] += eq_r_cycle[j];
+                }
                 j += 1;
             }
-            result
+            local_arrays
         })
         .reduce(
-            || unsafe_allocate_zero_vec(K),
+            || vec![unsafe_allocate_zero_vec(K); d],
             |mut running, new| {
-                running
-                    .par_iter_mut()
-                    .zip(new.into_par_iter())
-                    .for_each(|(x, y)| *x += y);
+                running.par_iter_mut().zip(new.into_par_iter()).for_each(
+                    |(running_arr, new_arr)| {
+                        running_arr
+                            .par_iter_mut()
+                            .zip(new_arr.into_par_iter())
+                            .for_each(|(x, y)| *x += y);
+                    },
+                );
                 running
             },
-        );
+        )
+        .into_iter()
+        .enumerate()
+        .for_each(|(i, arr)| F_arrays[i] = arr);
```

Comment:
instead of doing this might as well just initialize `F_arrays` to the result of the `reduce`